### PR TITLE
server: usage caps

### PR DESCRIPTION
**Pull request by raayan**
**From the CLI** ft. brutal logs
```zsh
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ lipsate 100000 | ../../target/debug/lockbook edit nice/test.txt --stdin
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ ../../target/debug/lockbook sync                                       
Pushing: test.txt
Sync complete.
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ lipsate 100000 | ../../target/debug/lockbook edit nice/test.txt --stdin
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ ../../target/debug/lockbook get-usage                                  
Uncompressed File Size: 100 KB
Server Utilization: 23.03 KB
Server Data Cap: 100 KB
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ ../../target/debug/lockbook sync                                       
Pushing: test.txt
Sync complete.
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ ../../target/debug/lockbook get-usage
Uncompressed File Size: 100 KB
Server Utilization: 22.46 KB
Server Data Cap: 100 KB
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ lipsate 500000 | ../../target/debug/lockbook edit nice/test.txt --stdin
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ ../../target/debug/lockbook get-usage                                  
Uncompressed File Size: 500 KB
Server Utilization: 22.46 KB
Server Data Cap: 100 KB
raayan@raayan-p15 ~/dev/lockbook/clients/cli [raayan/usage_caps] $ ../../target/debug/lockbook sync                    
Pushing: test.txt
error: Data cap exceeded! You'll need to upgrade your storage plan or reduce usage to continue syncing.
```

**From the server**
```zsh
[2021-07-11 15:40:39] [lockbook_server_lib::file_index_repo    ] [WARN ]: tried to sync fec0ae6c-eaa9-4783-bf57-ff508fd91750 old_size=22743 new_size=104722 cap=100000 current_usage=22743 projected_usage=104722
```

### CODE REVIEW

**Review by Parth — CHANGES_REQUESTED** [review]
Good first steps, we want this error to make it to `lib.rs` and be an explicit error for `sync`.

**Comment by raayan** [comment]
@Parth isn't @tvanderstad changing sync significantly atm?

**Comment by Parth** [comment]
Yeah he is, but this error will exist in his work too. I think there's only a small amount of stuff that will be duplicated. @tvanderstad a lot of errors are treated like: "pull and we'll retry", this is not one of those errors. So for both SyncV2 (what @raayan will slightly modify) and SyncV3 (what @tvanderstad is working on), if we see this error, we want to put the breaks on and propagate this up to the user.

imo this is a small part of this body of work, the more significant part is the UX around approaching the limit, warning users, and then finally indicating to them that it's time to buy more space.

If you wanted to pause this body of work until SyncV3 is merged, you could start working on some of the streams of work required for actually adding a payment method on various platforms (see the doc).

File: core/src/service/sync_service.rs [hunk 1]
```diff
@@ -123,6 +125,7 @@ pub fn sync(config: &Config, f: Option<Box<dyn Fn(SyncProgress)>>) -> Result<(),
                     debug!("{:#?} executed successfully", work_unit);
                     sync_errors.remove(&work_unit.get_metadata().id);
                 }
+                Err(CoreError::DataCapExceeded) => return Err(CoreError::DataCapExceeded),
```
Comment (raayan) [inline core/src/service/sync_service.rs:128]: This is notable, we boot out of the sync process here

**Comment by Parth** [comment]
> @Parth isn't @tvanderstad changing sync significantly atm?
@raayan was right

```
Conflicting files
clients/cli/src/edit.rs
clients/cli/src/main.rs
clients/cli/src/new.rs
clients/cli/src/utils.rs
core/libs/models/src/api.rs
core/src/lib.rs
core/src/service/sync_service.rs
server/server/src/file_index_repo.rs
server/server/src/file_service.rs
```

**Comment by Parth** [comment]
Will be copying from this PR in a new PR soon to be linked

**Pull request closed without merging**